Merge all Hubble server-side code into Cilium

tgraf opened the pull request:
This imports the server-side code of hubble to avoid depending on the
cilium/hubble repository. It does so with minimal code changes. It does
not remove unnecessary code yet.

Changes:
 - Fix all import paths
 - Replace gogo/protobuf with golang/protobuf
 - Annotate all tests with Cilium build flags
 - Replace use of sync.Mutex with lock.Mutex

gandro reviewed (commented):
This is going to be great! A couple thoughts on package organization going forward (some of this can also be done in future PRs once this is merged):

- Where will the "standalone" Hubble server code (i.e. the `hubble serve` command plus Dockerfile for the DaemonSet) live? This PR imports all the code necessary to implement it (e.g. `pkg/hubble/ipcache`, `pkg/hubble/servicecache`), without importing the actual binary. I personally think we should keep all support code related to stand-alone mode out of the Cilium repository.

- For the Cilium API, we have the generated files in `/api/v1` and custom (non-generated) code lives in `/pkg/api/v1`. For the Hubble API, we now also have the generated files in `/api/v1`, but the custom code on `pkg/hubble/api/v1`. In my opinion, we should clean up `pkg/hubble/api/v1` (it has a ton of random definitions) and then move the remaining type definitions in to`pkg/api/v1/hubble` instead.

rolinh commented on daemon/cmd/hubble.go:
```diff
@@ -18,17 +18,17 @@ import (
+	"github.com/cilium/cilium/pkg/hubble/observer"
+	"github.com/cilium/cilium/pkg/hubble/parser"
+	hubbleServer "github.com/cilium/cilium/pkg/hubble/server"
+	"github.com/cilium/cilium/pkg/hubble/server/serveroption"
```
This should be moved under `pkg/hubble/observer`.

rolinh commented on pkg/hubble/api/socket_test.go:
```diff
@@ -0,0 +1,38 @@
+// Copyright 2019 Authors of Hubble
```
Shouldn't headers be updated to something like the following?
```
Copyright 2019 Authors of Hubble
Copyright 2020 Authors of Cilium
```

tgraf replied:
Probably, this PR doesn't really change any code so I've just left copyright notices intact for now.

glibsm commented on pkg/hubble/observer/option/option.go:
```diff
@@ -12,12 +12,14 @@
 // limitations under the License.
 
-package serveroption
+package option
```
why this change?

tgraf replied:
I was on requested by @Rolinh to prepare for the peer API. I'll drop the commit for now.

tgraf committed:
hubble: Import server-side Hubble code

This imports the server-side code of hubble to avoid depending on the
cilium/hubble repository. It does so with minimal code changes. It does
not remove unnecessary code yet.

Changes:
 - Fix all import paths
 - Replace gogo/protobuf with golang/protobuf
 - Annotate all tests with Cilium build flags
 - Replace use of sync.Mutex with lock.Mutex

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
hubble: Remove unused code

This removes all code which is not required when Hubble runs in embedded
mode.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tklauser reviewed (approved):
Nice improvement on the binary size!

```
== daemon/cilium-agent ==
bss                                7813984     7812704       -1280
data                                684624      677360       -7264
text                              56297258    55761833     -535425
```
